'what is an image?' episode (#26)
HealthBioscienceIDEAS/microscopy-novice: episodes/what-is-an-image.md

## Pixels

Let's start by removing all layers from the Napari viewer and opening a new 
sample image:

- Click on the top layer in the layer list and shift + click the bottom layer. 
This should highlight all layers in blue.

- Press the remove layer button ![](
https://raw.githubusercontent.com/napari/napari/main/napari/resources/icons/delete.svg
){alt="A screenshot of Napari's delete layer button" height='30px'} 

- Go to the top menu-bar of Napari and select:  
`File > Open Sample > napari builtins > Human Mitosis`

![](fig/human-mitosis-napari.png){alt="A screenshot of a 2D image of human cells 
undergoing mitosis in Napari"}

This 2D image shows the nuclei of human cells undergoing mitosis. If we zoom in 
by scrolling, we can see that this image is actually made up of many small 
squares with different brightness. These squares are the image's pixels (or 
'picture elements') and are the individual units that make up all 
digital images.

If we hover over these pixels with the mouse cursor, we can see that each pixel 
has a specific value. Try hovering over pixels in dark and bright areas of the 
image and see how the value changes in the bottom left of the viewer:

> ![](fig/pixel-value.png){alt="A screenshot of Napari - with the mouse cursor 

Author: This would be useful! But I think it's a bit too much work for now - something to bear in mind if we have time in future

hovering over a pixel and highlighting the corresponding pixel value"}

You should see that brighter areas have higher values than darker areas (we'll 
see exactly how these values are converted to colours in the next episode).

## Images are arrays of numbers

We've seen that images are made of individual units called pixels that have 
specific values - but how is an image really represented in the computer? Let's 
dig deeper into Napari's `Image` layers...

First, open Napari's built-in Python console by pressing the console button ![](
https://raw.githubusercontent.com/napari/napari/main/napari/resources/icons/console.svg
){alt="A screenshot of Napari's console button" height='30px'}. 
Note this can take a few seconds to open, so give it some time:

![](fig/console.png){alt="A screenshot of Napari's console"}

:::::::::::::::::::::::::::::::::::::: callout

### Console readability

You can increase the font size in the console by clicking inside it, then 

> pressing Ctrl and + together. The font size can also be decreased with 

Reviewer: ```suggestion
pressing <kbd>Ctrl</kbd> and <kbd>+</kbd> together. The font size can also be decreased with 
```

> Ctrl and - together.

Reviewer: ```suggestion
<kbd>Ctrl</kbd> and <kbd>-</kbd> together.
```

Note that you can also pop the console out into its own window by clicking the 
small ![](
https://raw.githubusercontent.com/napari/napari/main/napari/resources/icons/pop_out.svg
){alt="A screenshot of Napari's float panel button" height='30px'} 
icon on the left side.

::::::::::::::::::::::::::::::::::::::::::::::::

Let's look at the human mitosis image more closely - copy the following into 

> Napari's console:

Reviewer: If this lesson is taught... how are we going to tell to the learners to "copy" this text? Will that be a code-along or shared via the codimd/hackmd/etherpad document?

> Napari's console:

Author: Good question! I think since the code to run in this lesson is pretty short + simple, it should be fine for people to code-along. The code in the 'dimensions and data types' exercise is a bit longer though, so probably worth sharing that in a document, so it can be directly copied and pasted

```python

# Get the image data for the first layer in Napari
image = viewer.layers[0].data

# Print the image values and type
print(image)

> print(type(image))

Reviewer: ~It may be useful to print the shape so they can see it's the same than what they see in the GUI (though with one value less)~
```suggestion
print(type(image))
print(image.shape)
```

Oh! you've got that later on! :sweat_smile:

image is created will depend on the type of microscope you are using e.g. 
widefield, confocal, superresolution etc. In general though, we have 3 main 
parts: 

- **Sample:** the object we want to image e.g. some cells
- **Objective lens:** the lens that gathers the light and focuses it for 
detection
- **Detector:** the device that detects the light to form the digital image 
e.g. a CCD camera

To briefly summarise, light is emitted from the sample (e.g. from a fluorescent 
marker), then gathered and focused with the objective lens, before hitting the 
detector. The detector is divided into many small, light sensitive areas - these 
are physical pixels, that will correspond to the pixels in the final image. When 
light hits one of these areas it is converted to one or more electrons, with 
more light resulting in more electrons and a higher final value for that pixel.

The important factor to understand is that the final pixel value is only ever an 
approximation of the real sample. Many factors will affect this final result 
including the microscope optics, detector performance etc.

::::::::::::::::::::::::::::::::::::: discussion 

> Read the ['A simple microscope' section of Pete Bankhead's bioimage book](

Reviewer 2: This book looks a really good resource. It makes me wonder how best to use this. Is the idea to make it as recommended reading ahead of the course? Or is the idea that we will stop for a few minutes while they read this section and then have a discussion?

Author: I was more thinking the latter - stopping for a few minutes to read this section + discuss. I think it's a bit large to have as recommended reading before the course, but I'll definitely link to it at the end as further reading - as it really is an excellent resource! I've started an issue to collect useful links/resources for the end: https://github.com/HealthBioscienceIDEAS/microscopy-novice/issues/29

The type determines what kind of values can be stored in the array, for example:

- Unsigned integer: positive whole numbers
- Signed integer: positive and negative whole numbers
- Float: positive and negative numbers with a decimal point e.g. 3.14

For our mitosis image, 'unsigned integer' means that only positive whole numbers 
can be stored inside. You can see this by hovering over the pixels in the image 
again in Napari - the pixel value down in the bottom left is always a positive 
whole number.

## Bit depth

The bit depth determines the range of values that can be stored e.g. only values 
between 0 and 255. This is directly related to how the array is stored in the 
computer.

In the computer, each pixel value will ultimately be stored in some binary 
format as a series of ones and zeros. Each of these ones or zeros is known 
as a 'bit', and the 'bit depth' is the number of bits used to store each value. 
For example, our mitosis image uses 8 bits to store each value (i.e. 
a series of 8 ones or zeros like `00000000`, or `01101101`...).

The reason the bit depth is so important is that it dictates the number of 
different values that can be stored. In fact it is equal to:

> \[\large \text{Number of values} = 2^{\text{bit depth}}\]

Reviewer 2: Maybe put bit depth in brackets

> Going back to our mitosis image, this means that it can store $2^8 = 256$ 

Reviewer 2: ```suggestion
Going back to our mitosis image, since it is stored with an integer with a bit-depth of 8, this means that it can store $2^8 = 256$ 
```

different values, which is equal to a range of 0-255 for unsigned integers.

We can verify this by looking at the maximum value of the mitosis image:

```python

print(image.max())
```

```output

255
```

You can also see this by hovering over the brightest nuclei in the viewer and 
examining their pixel values. Even the brightest nuclei won't exceed the limit 

> of 255.

Reviewer 2: I could also see you having them check the minimum and finding that it is 7, illustrating that even though a lot of the pixels look black none of them are 0 and they all have some low level signal/noise within them.

::::::::::::::::::::::::::::::::::::: challenge 

## Dimensions and data types

> Let's open a new image by copying and pasting the following lines into the 

Reviewer 2: Do you want them to remove the human mitosis layer first?

```

### Type and bit depth

The image's type and bit depth are: unsigned integer 16-bit

```python

image.dtype
```

```output

dtype('uint16')
```

### Min and max

Based on a bit depth of 16, this image can store $2^{16} = 65536$ values. As it 
is of type 'unsigned integer' this corresponds to a min and max of 0 and 65535.

:::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::

## Common data types

NumPy supports a [very wide range of data types](
https://numpy.org/doc/stable/reference/arrays.scalars.html#sized-aliases), 
but there are a few that are most common for image data:

| NumPy datatype     | Full name                |
| :---------         | :--------------          |
| `uint8`            | Unsigned integer 8-bit   |
| `uint16`           | Unsigned integer 16-bit  |
| `float32`          | Float 32-bit             |
| `float64`          | Float 64-bit             |

`uint8` and `uint16` are most common for images from light microscopes. 
`float32` and `float64` are common during image processing (as we will see in 
later episodes).

## Choosing a bit depth

Most images are either 8-bit or 16-bit - so how to choose which to use? 

> A higher bit depth will allow a wider range of values to be stored, but also 
> result in larger image files overall. In general, a 16-bit image will have a 

Reviewer 2: ```suggestion
A higher bit depth will allow a wider range of values to be stored, but it will also 
result in larger file sizes for the resulting images. In general, a 16-bit image will have a 
```

> file size about 2x as large as 8-bit without compression (we'll discuss 

Reviewer 2: ```suggestion
file size that is about twice as large as an 8-bit image if no compression is used (we'll discuss 
```

compression in a later episode).

The best bit depth choice will depend on your particular imaging experiment and 
research question. For example, if you know you have to recognise features that 
only differ slightly in their brightness, then you will likely need 16-bit to 
capture this. Equally, if you know that you will need to collect a very large 
number of images and 8-bit is sufficient to see your features of interest, then 
8-bit may be a better choice to reduce the required file storage space. As 
always it's about choosing the best fit for your specific project! 

For more information on bit depths and types - we highly recommend 

> [this chapter of Pete Bankhead's free bioimage book](

Reviewer: A more explicit link, rename it as appropriate.
```suggestion
[the chapter on bit depths of Pete Bankhead's free bioimage book](
```

https://bioimagebook.github.io/chapters/1-concepts/3-bit_depths/bit_depths.html).

:::::::::::::::::::::::::::::::::::::: callout

### Clipping and overflow

It's important to be aware of what image type and bit depth you are using. If 

> you try to store values outside of the valid range this can lead to clipping 
> and overflow.

Reviewer 2: ```suggestion
you try to store values outside of the valid range, this can lead to _clipping_ 
and _overflow_.
```

system is Napari using?

To look into this, let's hover over pixels in our mitosis image and examine the 
coordinates that appear to the left of the pixel value:

![](fig/coordinates.png){alt="A screenshot of Napari - with the mouse cursor 
hovering over a pixel and highlighting the corresponding coordinates"}

As you move around, you should see that the lowest coordinate values are at the 
top left corner, with the first value increasing as you move down and the second 
value increasing as you move to the right. This is different to the standard 
coordinate systems you may be used to (for example, from making graphs):

![](fig/coordinate-system.png){alt="Diagram comparing a standard graph 
coordinate system (left) and the image coordinate system (right)"}

Note that Napari lists coordinates as [y, x] or [rows, columns], so e.g. [1,3] 
would be the pixel in row 1 and column 3. Remember that these coordinates always 
start from 0 as you can see in the diagram below:

![](fig/coordinates-on-image.png
){alt="A diagram showing how pixel coordinates change over a simple 4x4 image" width=50%}

For the mitosis image, these coordinates are in pixels, but we'll see in later 

> episodes that images can also be scaled based on resolution (e.g. in 

Reviewer 2: ```suggestion
episodes that images can also be scaled based on resolution to represent distances in the physical world (e.g. in 
```

micrometres). Also, bear in mind that images with more dimensions (e.g. a 3D 

> image) will have longer coordinates like [z, y, x]...

Reviewer 2: I thought slice was typically stored in the third dimension so it would be [y,x,z] ?

Reviewer: I knot that python does it as `z, y, x` but different instruments may have axis stored in different orders. One of the plugin we developed had to manage these changes because the underlying C-code had been written in the axis as you said.

Author: Yes python doesn't have very strong conventions for the order of axes, it varies by library/plugin/filetype... It tends to be (z, y, x), but I've added a lot of coverage of this to the 'Multi-dimensional images' episode

2. What is the pixel value at x=25 and y=63?
3. Change the value of the pixel at x=10 and y=15 to 200. 
Check the new value - is it correct? If not, why not?
4. Change the value of the pixel at x=10 and y=15 to 300. 
Check the new value - is it correct? If not, why not?

:::::::::::::::::::::::: solution 

### Solution

### 1

```python

image[115, 213]
```

```output

162
```

### 2

```python

image[63, 25]
```

```output

9
```

### 3

```python

image[15, 10] = 200
viewer.layers["human_mitosis"].refresh()

print(image[15, 10])
```

```output

200
```

The new value is correct. If you zoom into the top left corner of the image, 
you should see the one bright pixel you just created.

### 4

```python

image[15, 10] = 300
viewer.layers["human_mitosis"].refresh()

print(image[15, 10])
```

```output

44
```

The new value is not correct. This is because 300 exceeds the maximum value for 
this 8-bit image (max 255). The value therefore overflows and 'wraps around' to 
give 44 - an incorrect value.

:::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: keypoints 

- Digital images are made of pixels

> - Digital images are arrays of numbers

Reviewer 2: ```suggestion
- Digital images stores these pixes as arrays of numbers
```

---
title: 'What is an image?'
teaching: 30
exercises: 10
---

:::::::::::::::::::::::::::::::::::::: questions 

- How are images represented in the computer?

::::::::::::::::::::::::::::::::::::::::::::::::

::::::::::::::::::::::::::::::::::::: objectives

- Explain how a digital image is made of pixels

- Find the value of different pixels in an image in Napari

> - Determine an image's dimensions (numpy ndarray .shape)
> 
> - Determine an image's data type (numy ndarray .dtype)

Reviewer: ```suggestion
- Determine an image's dimensions (numpy ndarray `.shape`)

- Determine an image's data type (numpy ndarray `.dtype`)
```
